pandas/_libs/tslib.pyx

```diff
@@ -301,6 +301,9 @@ def array_with_unit_to_datetime(
             iresult = values.astype("i8", copy=False)
             # fill missing values by comparing to NPY_NAT
             mask = iresult == NPY_NAT
+            # also handle NaN
+            if values.dtype.kind == "f":
+                mask |= values != values
```

Comment:
on the relevant hardware, what do you get for iresult when values is [np.nan]?

Reply:
$ python3
Python 3.10.7 (main, Sep 11 2022, 08:41:56) [GCC] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy as np
>>> np.array([np.nan]).astype("i8")
array([9223372036854775807])

Reply:
@seberg is getting 9223372036854775807 instead of 9223372036854775808 weird or is it just me?

Reply:
42 would also be a valid result.

Reply:
NumPy has no guarantees for `NaN -> int64` casts currently! (Although they are better implemented, in a sense).  So what you get is completely off-bets.  For example on arm, you may get 0!
And that breaks `mask = iresult == NPY_NAT`?  There should be tests failing here I would think?

However, NumPy *does* have guarantees for `NaN -> datetime/timedelta` casts, so going via `i8` seems dubious in this case?  The cast triggers undefined behavior in C.

I am not sure what you mean with:
> 42 would also be a valid result.

@andreas-schwab?

Reply:
No sure why you ask, you already gave the answer above.

Reply:
Misread the code.  In any case, NumPy handles the cast correctly for datetime64, even if it uses a slower/weirder code path (currently, this is very much fixable!).  So the question is wether it doesn't make more sense to cast to datetime64.

Further, the above may cause floating-point warnings in newer NumPy versions, which is something you probably don't want to see!

Reply:
this is fine.  can you add a comment pointing back to this PR and a few words about why we're doing this